core/state/state_object.go

```diff
@@ -358,37 +377,47 @@ func (s *stateObject) updateTrie(db Database) Trie {
 	if len(s.pendingStorage) > 0 {
 		s.pendingStorage = make(Storage)
 	}
-	return tr
+	return tr, nil
 }
 
-// UpdateRoot sets the trie root to the current root hash of
+// UpdateRoot sets the trie root to the current root hash of. An error
+// will be returned if trie root hash is not computed correctly.
 func (s *stateObject) updateRoot(db Database) {
+	tr, err := s.updateTrie(db)
+	if err != nil {
+		s.setError(fmt.Errorf("updateRoot (%x) error: %v", s.address, err))
+		return
+	}
 	// If nothing changed, don't bother with hashing anything
-	if s.updateTrie(db) == nil {
+	if tr == nil {
 		return
 	}
 	// Track the amount of time wasted on hashing the storage trie
 	if metrics.EnabledExpensive {
 		defer func(start time.Time) { s.db.StorageHashes += time.Since(start) }(time.Now())
 	}
-	s.data.Root = s.trie.Hash()
+	s.data.Root = tr.Hash()
```

Comment:
This is an interesting change. One one hand, it's somewhat logical, less data access. On the other hand, previously we implicitly checked if `s.trie` was actually set to the newly created/update/returned `tr`. With the change, now we do not check this side effect. 

On the other hand, the original code has no guarantee either that `s.trie` and `tr` is actually the same.

Soo, I guess my question is, is there a specific reason for `updateTrie` returning the Trie, instead of say a boolean flag to signal no update (the current `nil` return). If we were to change the update to just return a flag, there would be no duality of two potentially conflicting ways to access the - hopefully - same trie.

Reply:
I will keep the current behavior.

StateDB is really complicated :) 

In commitTrie commit operation will be skipped if trie is nil. Commit only happens at the end of block(post-byzantium), 
it can happen the last tx in block doesn't touch this storage trie, but this storage trie is indeed dirty. 

If we simply use a boolean, it will lead data be missing. `tr == nil` magically represents this tris is not updated in the entire block :)

I agree we need to improve it. Give me a bit more time to figure out a correct solution.

```go
// commitTrie submits the storage changes into the storage trie and re-computes
// the root. Besides, all trie changes will be collected in a nodeset and returned.
func (s *stateObject) commitTrie(db Database) (*trie.NodeSet, error) {
	tr, err := s.updateTrie(db)
	...
	// If nothing changed, don't bother with committing anything
	if tr == nil {
		return nil, nil
	}
	root, nodes, err := tr.Commit(false)
	if err == nil {
		s.data.Root = root
	}
	return nodes, err
}
```